lib/stdlib/src/proc_lib.erl

```diff
@@ -331,60 +320,88 @@ start_link(M, F, A) when is_atom(M), is_atom(F), is_list(A) ->
-    Pid = ?MODULE:spawn_link(M, F, A),
-    sync_wait(Pid, Timeout).
+    sync_start_link(?MODULE:spawn_link(M, F, A), Timeout).
 
 -spec start_link(Module, Function, Args, Time, SpawnOpts) -> Ret when
       Module :: module(),
       Function :: atom(),
       Args :: [term()],
       Time :: timeout(),
-      SpawnOpts :: [spawn_option()],
+      SpawnOpts :: [start_spawn_option()],
       Ret :: term() | {error, Reason :: term()}.
 
 start_link(M,F,A,Timeout,SpawnOpts) when is_atom(M), is_atom(F), is_list(A) ->
-    Pid = ?MODULE:spawn_opt(M, F, A, ensure_link(SpawnOpts)),
-    sync_wait(Pid, Timeout).
+    ?VERIFY_NO_MONITOR_OPT(M, F, A, Timeout, SpawnOpts),
+    sync_start_link(?MODULE:spawn_opt(M, F, A, [link|SpawnOpts]), Timeout).
 
-sync_wait(Pid, Timeout) ->
+sync_start_link(Pid, Timeout) ->
     receive
 	{ack, Pid, Return} ->
-	    Return;
+            Return;
 	{'EXIT', Pid, Reason} ->
-	    {error, Reason}
+            {error, Reason}
     after Timeout ->
-	    unlink(Pid),
-	    exit(Pid, kill),
-	    flush(Pid),
-	    {error, timeout}
+            kill_flush(Pid),
+            {error, timeout}
     end.
 
-sync_wait_mon({Pid, Ref}, Timeout) ->
+-spec start_monitor(Module, Function, Args) -> {Ret, Mon} when
+      Module :: module(),
+      Function :: atom(),
+      Args :: [term()],
+      Mon :: reference(),
+      Ret :: term() | {error, Reason :: term()}.
+
+start_monitor(M, F, A) when is_atom(M), is_atom(F), is_list(A) ->
+    start_monitor(M, F, A, infinity).
+
+-spec start_monitor(Module, Function, Args, Time) -> {Ret, Mon} when
+      Module :: module(),
+      Function :: atom(),
+      Args :: [term()],
+      Time :: timeout(),
+      Mon :: reference(),
+      Ret :: term() | {error, Reason :: term()}.
+
+start_monitor(M, F, A, Timeout) when is_atom(M), is_atom(F), is_list(A) ->
+    sync_start_monitor(spawn_mon(M, F, A), Timeout).
+
+-spec start_monitor(Module, Function, Args, Time, SpawnOpts) -> {Ret, Mon} when
+      Module :: module(),
+      Function :: atom(),
+      Args :: [term()],
+      Time :: timeout(),
+      SpawnOpts :: [start_spawn_option()],
+      Mon :: reference(),
+      Ret :: term() | {error, Reason :: term()}.
+
+start_monitor(M,F,A,Timeout,SpawnOpts) when is_atom(M),
+                                            is_atom(F),
+                                            is_list(A) ->
+    ?VERIFY_NO_MONITOR_OPT(M, F, A, Timeout, SpawnOpts),
+    sync_start_monitor(?MODULE:spawn_opt(M, F, A, [monitor|SpawnOpts]),
+                       Timeout).
+
+sync_start_monitor({Pid, Ref}, Timeout) ->
     receive
 	{ack, Pid, Return} ->
-	    erlang:demonitor(Ref, [flush]),
-	    Return;
-	{'DOWN', Ref, _Type, Pid, Reason} ->
-	    {error, Reason};
-	{'EXIT', Pid, Reason} -> %% link as spawn_opt?
-	    erlang:demonitor(Ref, [flush]),
-	    {error, Reason}
+            {Return, Ref};
+	{'DOWN', Ref, process, Pid, Reason} = Down ->
+            self() ! Down,
```

Comment:
This was actually my first approach, but the return type got really ugly and I think it got harder for the user to know what to expect. Now it is easy to know whether you will get a `DOWN` message or not, since you always get a `DOWN` if the child terminates. In the `{error, timeout}` case you can also choose if you want to wait for the `DOWN` to reach you or not. By waiting for it you know that the child has terminated and released all resources, which you do not know if you don't wait.

In the `gen_[server|statem|event]:start_monitor()` you do not get a `DOWN` messages in the `error` cases. This is however an easier scenario, since all return values already have been defined which is not the case for `proc_lib:start_monitor()`.